include/dem/from_file_insertion.h

```diff
@@ -0,0 +1,108 @@
+
+template <int dim>
+class FromFileInsertion : public Insertion<dim>
+{
+public:
+  FromFileInsertion(const DEMSolverParameters<dim> &dem_parameters,
+                    const std::vector<std::shared_ptr<Distribution>>
+                      &distribution_object_container);
+
+  /**
+   * @brief The FromFileInsertion class inserts particles using data stored in a file.
+   * This allows the insertion of any number of particles at a well-controlled
+   * location with any diameter value, translation and angular velocity.
+   *
+   * @param particle_handler The particle handler of particles which are being
+   * inserted
+   * @param triangulation Triangulation to access the cells in which the
+   * particles are inserted
+   * @param dem_parameters DEM parameters declared in the .prm file
+   */
+  virtual void
+  insert(Particles::ParticleHandler<dim>                 &particle_handler,
+         const parallel::distributed::Triangulation<dim> &triangulation,
+         const DEMSolverParameters<dim> &dem_parameters) override;
+
+
+  /**
+   * @brief Carries out assigning the properties of inserted particles specifically
+   * for the from_file insertion method. In this method, the initial translation
+   * and angular velocities and the diameter of each particles is set.
+   *
+   * @param dem_parameters DEM parameters declared in the .prm file
+   * @param inserted_this_step_this_proc Number of particles that are inserted
+   * at each insertion step on each processor. This value can change in the last
+   * insertion step to reach the desired number of particles
+   * @param particles_data Contains the particles
+   * @param particle_properties Properties of all inserted particles at each insertion step
+   */
+  void
+  assign_particle_properties_for_from_file_insertion(
+    const DEMSolverParameters<dim>             &dem_parameters,
+    const unsigned int                         &inserted_this_step_this_proc,
+    std::map<std::string, std::vector<double>> &particles_data,
+    std::vector<std::vector<double>>           &particle_properties);
+
+
+
+  /**
+   * @brief Serialize the list insertion object to an output archive.
+   *
+   * @param ar Output archive where the attributes are stored.
+   */
+  virtual void
+  serialize(boost::archive::text_oarchive &ar, const unsigned int) override
+  {
+    ar &remaining_particles_of_each_type &current_inserting_particle_type;
+  }
+
+  /**
+   * @brief Deserialize an input archive to the list insertion object.
+   *
+   * @param ar Input archive where the attributes are stored.
+   *
+   */
+  virtual void
+  deserialize(boost::archive::text_iarchive &ar, const unsigned int) override
+  {
+    ar &remaining_particles_of_each_type &current_inserting_particle_type;
+  }
+
+  // Number of remaining particles of each type that should be inserted in the
+  // upcoming insertion steps
+  unsigned int remaining_particles_of_each_type;
+  unsigned int current_inserting_particle_type;
```

Comment:
How does this parameter work ?

Reply:
current_insertion_particle_type is the "type", so basically always zero, except when we are using particles with different properties (Young's modulus, friction coefficient, etc)

remaining_particles_of_each_type is the remaining particle we have to insert of the current type. So, if we are using two types of particles (1000 of type 0 and 2000 of type 1) and there's all ready 250 of inserted particles of type 0, then  remaining_particles_of_each_type = 750. It will go back to 2000 when it will start to insert the type 1. 

Remember, we are inserting type 0, then type 1, etc. 

This follows the same structure has list insertion (and every other insertion method I think)